### PR TITLE
Fixes #207 

### LINKED ISSUE #207: Body kinematics when origin not at center of mass
When the origin of a body is not at the center of mass (including any attached objects), the kinematics of bodies are not correct.

As a somewhat extreme demonstration of this, we can set up two kinematically identical situations, one where the body is at the center of mass, and one where the body is far from the COM.

```
--------------------- MoorDyn Input File -------------------------------------------------------
Testing body kinematics when body origin != COM
---------------------- ROD TYPES ------------------------------------
TypeName      Diam     Mass/m    Cd     Ca      CdEnd    CaEnd
(name)        (m)      (kg/m)    (-)    (-)     (-)      (-)
buoy          1       1.0e1     1.2    0.0     1.2      0.0
---------------------------- BODIES -----------------------------------------------------
ID   Attachment  X0     Y0    Z0     r0      p0     y0     Mass  CG*   I*      Volume   CdA*   Ca
(#)     (-)      (m)    (m)   (m)   (deg)   (deg)  (deg)   (kg)  (m)  (kg-m^2)  (m^3)   (m^2)  (-)
1       free      0     0      20      0.0       0      0      0     0    0        0       0      0
2       free      0     10      20     0.0       0      0      0     0    0        0       0      0
---------------------- RODS ----------------------------------------
ID     RodType   Attachment  Xa    Ya    Za    Xb    Yb    Zb    NumSegs   RodOutputs
(#)    (name)    (#/key)     (m)   (m)   (m)   (m)   (m)   (m)   (-)       (-)
1      buoy     Body1         0     0     -2     0   0.0    2     10        -
2      buoy     Body2         0     -10   -2     0  -10.0   2     10        -
```

In this case we have the same 4 meter tall rod, with it's center at (0, 0, 20). 
Body 1 starts at (0, 0, 20), and body 2 starts 10 meters to the side in the +y direction (with the rod a corresponding -10 meters to the side to be back at the origin).

In this simulation, gravity is set to 0, and we give the two bodies velocities equivalent to their rod rotating at 0.2 rad/sec around the x axis. 
For body 1 this means an initial velocity vector of  
`0.0, 0.0, 0.0, 0.2, 0.0, 0.0` 
and for body 2 this means an initial velocity vector of  
`0.0, 0.0, 10 * 0.2, 0.2, 0.0, 0.0`.  
The 10*0.2 m/s velocity in the +z direction is due to the fact that the body is 10 meters from the center of rotation. 

Running this simulation and viewing the results in paraview this is what we see:

https://github.com/FloatingArrayDesign/MoorDyn/assets/129765925/585ad2e4-9520-4bb9-9e00-ce97854c2b29

This is obviously incorrect, since both situations should be equivalent in terms of rod position/rotation.

In terms of solutions, I don't have much beyond avoiding this sort of scenario. I'm pretty certain that the problem is related to missing non-inertial force on the body (likely at least the centrifugal force). But I don't have a good grasp on how that would need to be implemented. 
I will note that in more practical situations, things like fluid drag or buoyancy forces often do a pretty good job of mitigating most of the error caused by the incorrect kinematics.

Additional complications include
* Added mass turns masses from scalars to matrices makes deriving the necessary equations more complicated. 
* I've found very little literature on handling kinematics of an object in a global coordinate frame when the object origin is not at the center of mass. Most things I've found either uses a body fixed coordinate frame, or require the object origin to be at the center of mass.

I don't necessarily have a strong feeling about how much effort should be put into addressing this issue. If there was sufficient reference theory and desire from other, I likely could put a fair amount of time into this. But I figured I'd bring this up to determine if it was a known limitation, and if there had been any other work (perhaps in OpenFast) to handle issues like this.

### PR DESCRIPTION
**Pull request by sanguinariojoe**
See #207

### CODE REVIEW

**Comment by sanguinariojoe** [comment]
Hahaha, I started implementing it as omega x omega x..., and then I
thought, what a waste!

Sometimes I amuse myself... Anyway, you are totally right. I am correcting
that ASAP

File: source/Body.cpp [hunk 1]
```diff
@@ -588,6 +589,7 @@ Body::doRHS()
 		vec6 F6_i;
 		mat6 M6_i;
 		attached->getNetForceAndMass(F6_i, M6_i, r7.pos);
+		F6_i(Eigen::seqN(0, 3)) += attached->getCentripetalForce(r7.pos, v6);
```
Comment (RyanDavies19) [inline source/Body.cpp:592]: Should this be moved into getNetForceAndMass in the interest of keeping consistent function names? NetForce to me includes this centripetal force, and that is also where forces are translated to be about the body's ref point

**Comment by RyanDavies19** [comment]
@AlexWKinley @sanguinariojoe Thank you all for putting this together and working out the math. This all looks good on my end, and I agree that MD was failing to account for the non-inertial reference frame. I will talk with @RBergua about this n the MDF side and address it in a PR there. I'm curious @AlexWKinley does this resolve your surface piercing buoys issues? USFLOWT is using MoorDyn for the same purpose.

**Comment by AlexWKinley** [comment]
I think this will do a solid job addressing my issues. I haven't had a chance to particularly rigorously test this. I'll try to find some time in the next day or two.

The overall surface piercing behavior is probably more complicated than just this particular issue, but that can be a later investigation.

**Comment by RyanDavies19** [comment]
Once I get this in on the OpenFAST side I'll let you all know how it does resolving issues with USFLOWT. Agreed though the surface piercing is probably more than just this. If I come across any fixes on the MD-F side I'l open an issue here

File: source/Rod.cpp [hunk 2]
```diff
@@ -816,6 +816,20 @@ Rod::getNetForceAndMass(vec6& Fnet_out, mat6& M_out, vec rRef)
...
+vec
+Rod::getCentripetalForce(vec rRef, vec w) const
+{
+	if (!N)
+		return vec::Zero();
+
+	vec F = vec::Zero();
+	for (unsigned int i = 0; i <= N; i++) {
+		const vec rRel = r[i] - rRef;
+		F += w.squaredNorm() * (M[i] * rRel);
```
[the remark this answers is not in the document]
Comment (sanguinariojoe) [reply source/Rod.cpp:828]: Hahaha, I started implementing it as omega x omega x..., and then I
thought, what a waste!

Sometimes I amuse myself... Anyway, you are totally right. I am correcting
that ASAP

[the remark this answers is not in the document]
Comment (sanguinariojoe) [reply source/Rod.cpp]: Solved in 0838d46ac0acab882ac9e0a42f610898441cd3ab

File: source/Body.cpp (hunk 1 above)
Comment (sanguinariojoe) [reply source/Body.cpp]: Fixed in 94b78a27a35b47180730db5f3c8d47eac4726238

**Comment by AlexWKinley** [comment]
Alright, I've had some time to test out this PR. Overall, everything works as expected.

The only additional thought I have related to this PR, is that we might want to also handle when the body CG is defined as not being at the origin of the body. I think this should be pretty straight forward. The code below seems to work at least in the context where there isn't added mass.

```c++
	// First, the body's own mass matrix must be adjusted based on its
	// orientation so that we have a mass matrix in the global orientation frame
	M = rotateMass6(OrMat, M0);

	// gravity forces and moments about body ref point given CG location
	const vec body_rCGrotated = OrMat * body_rCG;
	// weight+buoyancy vector
	const vec Fgrav =
	    vec(0.0, 0.0, bodyV * env->rho_w * env->g - bodyM * env->g);
	F6net(Eigen::seqN(0, 3)) = Fgrav;
	F6net(Eigen::seqN(3, 3)) = body_rCGrotated.cross(Fgrav);

	// Centrifugal force due to COM not being at body origin
	const vec w = v6.tail<3>();
	F6net.head<3>() -=
	    M.topLeftCorner(3, 3) * (w.cross(w.cross(body_rCGrotated)));
```

As potential future investigation, I now think @sanguinariojoe is correct that our 6dof kinematics are missing some sort of $\omega \times I_{cm} \omega$ term. This come up if you give a body (with an attached rod), rotation around two separate axes. It does happen that in that case, the kinematics differ depending on where the body is, but both of the kinematics are incorrect so I don't think that should prevent merging this PR.

**Pull request merged**